dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLitmusVulkan.java

```diff
@@ -0,0 +1,294 @@
+        if (ctx != null) {
+            for (LitmusVulkanParser.SswDeclaratorContext sswDeclaratorContext : ctx.sswDeclarator()) {
+                int threadId0 = sswDeclaratorContext.threadId(0).id;
+                int threadId1 = sswDeclaratorContext.threadId(1).id;
+                programBuilder.addSwwPairThreads(threadId0, threadId1);
+            }
+        }
+        return null;
+    }
+
+    // ----------------------------------------------------------------------------------------------------------------
+    // Thread declarator list (on top of instructions)
+    @Override
+    public Object visitThreadDeclaratorList(LitmusVulkanParser.ThreadDeclaratorListContext ctx) {
+        for (LitmusVulkanParser.ThreadScopeContext threadScopeContext : ctx.threadScope()) {
+            int subgroupID = threadScopeContext.subgroupScope().scopeID().id;
+            int workgroupID = threadScopeContext.workgroupScope().scopeID().id;
+            int queuefamilyID = threadScopeContext.queuefamilyScope().scopeID().id;
+            // NB: the order of scopeIDs is important
+            programBuilder.newScopedThread(Arch.VULKAN, threadScopeContext.threadId().id,
+                    queuefamilyID, workgroupID, subgroupID);
+            threadCount++;
+        }
+        return null;
+    }
+
+    // ----------------------------------------------------------------------------------------------------------------
+    // Instruction list (the program itself)
+    @Override
+    public Object visitConstant(LitmusVulkanParser.ConstantContext ctx) {
+        return ExpressionFactory.getInstance().parseValue(ctx.getText(), TypeFactory.getInstance().getArchType());
+    }
+
+    @Override
+    public Object visitRegister(LitmusVulkanParser.RegisterContext ctx) {
+        return programBuilder.getOrNewRegister(mainThread, ctx.getText(), TypeFactory.getInstance().getArchType());
+    }
+
+    @Override
+    public Object visitInstructionRow(LitmusVulkanParser.InstructionRowContext ctx) {
+        for (int i = 0; i < threadCount; i++) {
+            mainThread = i;
+            visitInstruction(ctx.instruction(i));
+        }
+        return null;
+    }
+
+    @Override
+    public Object visitStoreConstant(LitmusVulkanParser.StoreConstantContext ctx) {
+        MemoryObject object = programBuilder.getOrNewMemoryObject(ctx.location().getText());
+        IConst constant = (IConst) ctx.constant().accept(this);
+        Boolean atomatic = ctx.atomatic().isAtomic;
+        String mo = ctx.mo().content;
+        String scope = ctx.scope().content;
+        String storageClass = ctx.storageClass().content;
+        String classSemantic = ctx.storageClassSemantic().content;
+        String avvisSemantic = ctx.avvisSemantic().content;
+        Store store = EventFactory.newStoreWithMo(object, constant, mo);
+        tagChecker(store, atomatic, mo, scope, avvisSemantic);
+        store.addTags(scope, storageClass, classSemantic);
```

Comment:
shouldn't you also add the `avvisSemantic` tag (at least [if the instruction is atomic](https://github.com/KhronosGroup/Vulkan-MemoryModel/blob/master/alloy/litmus.cpp#L191-L199))? It seems this comment apply to most of the visitors

Reply:
avvisSemantic is added inside the tagChecker (to check if `available` semantic goes with `release` mo and `visiable` semantic goes with `acquire` mo)

Reply:
I see `ATOM` being added at the end of that method, but I don't see `VIS` or `AV` being added, you just checked that av/vis is consistent with acq/rel, but you never add the av/vis tags

Reply:
I was thinkng `av`/`vis` as cases of `mo` and they will be added [here](https://github.com/tonghaining/Dat3M/blob/vulkan/dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLitmusVulkan.java#L295)

Reply:
As a general comment ... I have seen you treat `VIS` and `AV` as `mo` (memory order) ... while I suspect this is not really causing problem, it seems weird from a modeling / semantic perspective. The memory order is something that is common to most architectures we support (others use fences to guarantee memory orders). However `AV` and `VIS` are simply tags

Reply:
Okay I thought they were also `mo`s, I'll change them to be seperately tags in our grammar then